vm/systemSmartContracts/auction.go

```diff
@@ -0,0 +1,694 @@
+type StakedData struct {
+	StartNonce    uint64 `json:"StartNonce"`
+	Staked        bool   `json:"Staked"`
+	UnStakedNonce uint64 `json:"UnStakedNonce"`
+	UnStakedEpoch uint32 `json:"UnStakedEpoch"`
+	RewardAddress []byte `json:"RewardAddress"`
+}
+
+// AuctionConfig represents the settings for a specific epoch
+type AuctionConfig struct {
+	MinStakeValue *big.Int `json:"MinStakeValue"`
+	NumNodes      uint32   `json:"NumNodes"`
+	TotalSupply   *big.Int `json:"TotalSupply"`
+	MinStep       *big.Int `json:"MinStep"`
+	NodePrice     *big.Int `json:"NodePrice"`
+}
+
+type stakingAuctionSC struct {
+	eei           vm.SystemEI
+	unBoundPeriod uint64
+	kg            crypto.KeyGenerator
+	baseConfig    AuctionConfig
+}
+
+// NewStakingAuctionSmartContract creates an auction smart contract
+func NewStakingAuctionSmartContract(
+	minStakeValue *big.Int,
+	minStepValue *big.Int,
+	totalSupply *big.Int,
+	unBoundPeriod uint64,
+	numNodes uint32,
+	eei vm.SystemEI,
+	kg crypto.KeyGenerator,
+) (*stakingAuctionSC, error) {
+	if minStakeValue == nil {
+		return nil, vm.ErrNilInitialStakeValue
+	}
+	if minStakeValue.Cmp(big.NewInt(0)) < 1 {
+		return nil, vm.ErrNegativeInitialStakeValue
+	}
+	if check.IfNil(eei) {
+		return nil, vm.ErrNilSystemEnvironmentInterface
+	}
+
+	baseConfig := AuctionConfig{
+		MinStakeValue: minStakeValue,
+		NumNodes:      numNodes,
+		TotalSupply:   totalSupply,
+		MinStep:       minStepValue,
+		NodePrice:     big.NewInt(0).Set(minStakeValue),
+	}
+
+	reg := &stakingAuctionSC{
+		eei:           eei,
+		unBoundPeriod: unBoundPeriod,
+		kg:            kg,
+		baseConfig:    baseConfig,
+	}
+	return reg, nil
+}
+
+// Execute calls one of the functions from the staking smart contract and runs the code according to the input
+func (s *stakingAuctionSC) Execute(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if CheckIfNil(args) != nil {
+		return vmcommon.UserError
+	}
+
+	switch args.Function {
+	case "_init":
+		return s.init(args)
+	case "stake":
+		return s.stake(args)
+	case "unStake":
+		return s.unStake(args)
+	case "unBound":
+		return s.unBound(args)
+	case "claim":
+		return s.claim(args)
+	case "slash":
+		return s.slash(args)
+	case "get":
+		return s.get(args)
+	}
+
+	return vmcommon.UserError
+}
+
+func (s *stakingAuctionSC) get(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	if len(args.Arguments) < 1 {
+		return vmcommon.UserError
+	}
+
+	value := s.eei.GetStorage(args.Arguments[0])
+	s.eei.Finish(value)
+
+	return vmcommon.Ok
+}
+
+func (s *stakingAuctionSC) setConfig(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
+	ownerAddress := s.eei.GetStorage([]byte(ownerKey))
+	if !bytes.Equal(ownerAddress, args.CallerAddr) {
+		log.Debug("setConfig function called by not the owners address")
+		return vmcommon.UserError
+	}
+
+	if len(args.Arguments) != 6 {
+		log.Debug("setConfig function called with wrong number of arguments")
+		return vmcommon.UserError
```

Comment:
receipts are in place for VM. we need to change the output structure to return error string instead of error code. Error code is saved in the smart contract result, so error string would be saved and you know the error.

Reply:
Maybe error instead of error string for easier testing procedures.